File: C/Hacks/Hack06/C/colorUtils.c
#include "colorUtils.h"
#include <math.h>

/*
 * Author: Anthony Candelas
 * Date: 09OCT2025
 * Email: tsallinger2@unl.edu
 *
 * Description: Implements color transformation functions including
 * grayscale, sepia, and RGB↔CMYK conversions. Handles pass-by-reference
 * operations, pointer validation, and error conditions.
 */

int rgbToCMYK(int r, int g, int b, double *c, double *m, double *y, double *k) {
    if (!c || !m || !y || !k) return 1; // Null pointer error
    if (r < 0 || r > 255 || g < 0 || g > 255 || b < 0 || b > 255) return 2; // Out-of-range error

    double rPrime = r / 255.0;
    double gPrime = g / 255.0;
    double bPrime = b / 255.0;

    *k = 1.0 - fmax(fmax(rPrime, gPrime), bPrime);

    if (*k == 1.0) { // black edge case
        *c = *m = *y = 0.0;
        return 0;
    }

    *c = (1.0 - rPrime - *k) / (1.0 - *k);
    *m = (1.0 - gPrime - *k) / (1.0 - *k);
    *y = (1.0 - bPrime - *k) / (1.0 - *k);

    return 0;
}

int cmykToRGB(double c, double m, double y, double k, int *r, int *g, int *b) {
    if (!r || !g || !b) return 1; // Null pointer error
    if (c < 0 || c > 1 || m < 0 || m > 1 || y < 0 || y > 1 || k < 0 || k > 1) return 2; // Out-of-range error

    *r = (int)round(255.0 * (1.0 - c) * (1.0 - k));
    *g = (int)round(255.0 * (1.0 - m) * (1.0 - k));
    *b = (int)round(255.0 * (1.0 - y) * (1.0 - k));

    return 0;
}

void applyGrayscale(RGB *color) {
    if (color == NULL) return;

    int gray = (int)(0.3 * color->red + 0.59 * color->green + 0.11 * color->blue);
    color->red = gray;
    color->green = gray;
    color->blue = gray;
}

void applySepia(RGB *color) {
    if (color == NULL) return;

    int originalRed = color->red;
    int originalGreen = color->green;
    int originalBlue = color->blue;

    int sepiaRed = (int)(0.393 * originalRed + 0.769 * originalGreen + 0.189 * originalBlue);
    int sepiaGreen = (int)(0.349 * originalRed + 0.686 * originalGreen + 0.168 * originalBlue);
    int sepiaBlue = (int)(0.272 * originalRed + 0.534 * originalGreen + 0.131 * originalBlue);

    color->red = sepiaRed > 255 ? 255 : sepiaRed;
    color->green = sepiaGreen > 255 ? 255 : sepiaGreen;
    color->blue = sepiaBlue > 255 ? 255 : sepiaBlue;
}

void applyColorFilter(RGB *color, ColorType type) {
    if (color == NULL) return;

    switch (type) {
        case GRAYSCALE:
            applyGrayscale(color);
            break;
        case SEPIA:
            applySepia(color);
            break;
        default:
            break;
    }
}

File: Java/Hack06/ColorUtils.java
/*
 * Name: Anthony Candelas
 * Email: tsallinger2@unl.edu
 * Date: 03NOV2025
 * Lab/Task: Hack06 - Color Utilities
 *
 * Description:
 * Color utility methods for applying grayscale and sepia filters to RGB pixel
 * values. Functions operate on RGB objects and document behavior in the header.
 */

package unl.soc;

public class ColorUtils {

    /**
     * Enumeration representing available color filter types.
     */
    public enum ColorFilter {
        GRAYSCALE,
        SEPIA
    }

    /**
     * Represents an RGB color with public fields for red, green, and blue.
     * Values are expected to be in the range [0, 255].
     */
    public static class RGB {
        public int red;
        public int green;
        public int blue;

        /**
         * Constructs a new RGB object with the specified color components.
         *
         * @param red   red component (0–255)
         * @param green green component (0–255)
         * @param blue  blue component (0–255)
         */
        public RGB(int red, int green, int blue) {
            this.red = red;
            this.green = green;
            this.blue = blue;
        }
    }

    /**
     * Applies a specified color filter to the given RGB color.
     * The transformation is done in place (pass-by-reference).
     *
     * @param rgb     RGB color object to be modified
     * @param filter  color filter to apply (GRAYSCALE or SEPIA)
     */
    public static void applyFilter(RGB rgb, ColorFilter filter) {
        if (rgb == null || filter == null) {
            return;
        }

        switch (filter) {
            case GRAYSCALE:
                applyGrayscale(rgb);
                break;
            case SEPIA:
                applySepia(rgb);
                break;
        }
    }

    /**
     * Converts the given RGB color to grayscale using a weighted average formula.
     *
     * @param rgb RGB color object to convert to grayscale
     */
    public static void applyGrayscale(RGB rgb) {
        int gray = (int) Math.round(rgb.red * 0.299 + rgb.green * 0.587 + rgb.blue * 0.114);
        rgb.red = gray;
        rgb.green = gray;
        rgb.blue = gray;
    }

    /**
     * Converts the given RGB color to sepia tone using the standard sepia formula.
     *
     * @param rgb RGB color object to convert to sepia
     */
    public static void applySepia(RGB rgb) {
        int originalRed = rgb.red;
        int originalGreen = rgb.green;
        int originalBlue = rgb.blue;

        rgb.red = clamp((int) Math.round(0.393 * originalRed + 0.769 * originalGreen + 0.189 * originalBlue));
        rgb.green = clamp((int) Math.round(0.349 * originalRed + 0.686 * originalGreen + 0.168 * originalBlue));
        rgb.blue = clamp((int) Math.round(0.272 * originalRed + 0.534 * originalGreen + 0.131 * originalBlue));
    }

    /**
     * Ensures that a color value stays within the valid RGB range [0, 255].
     *
     * @param value the color component value
     * @return the clamped value
     */
    public static int clamp(int value) {
        return Math.min(255, Math.max(0, value));
    }
}